### PR TITLE
fetch analysis node query data when it is ready, and it is node on top

### PR DESCRIPTION
**Pull request by matallo**
from https://github.com/CartoDB/cartodb/issues/11874#issuecomment-292944611

- We should only request those models from the top node of the layer. It is not necessary to request all nodes data because we only "visualize" the data referenced to the last/top node.
- We must not request the models data at the beginning. We don't need it (although geometry info, but it is requested in other way), and it could provoke 3x the analyses the map has. 5 analyses x 3 objects (QSM, QGM, QRQ) = 15 requests :|

I made this a separate PR to properly track changes, I think the first point is achieved with no unintended consequences (as far as I tested). The second point seems to be tackled in the original PR.

Added the essential tests, and removed some failing that were not needed anymore.

Happy to discuss this further @xavijam

### CODE REVIEW

**Comment by matallo** [comment]
leaving the fetch as it was, moving the responsibility of is node on top to the collection, now just noticing the collectiong that the model is ready

/cc @xavijam @ivanmalagon

**Comment by javierarce** [comment]
👋  It seems that thew NEWS.md is not updated, @matallo.

File: app/views/admin/organization_users/_form.html.erb [hunk 1]
```diff
@@ -184,7 +184,7 @@
 
     <div class="FormAccount-row">
       <div class="FormAccount-rowLabel">
-        <label class="CDB-Text CDB-Size-medium is-semibold u-mainTextColor">Extra Here Isolines</label>
+        <label class="CDB-Text CDB-Size-medium is-semibold u-mainTextColor">Extra Routing API calls</label>
```
Comment (javierarce) [inline app/views/admin/organization_users/_form.html.erb]: Does this change belong to this issue?

File: lib/assets/core/javascripts/cartodb3/data/analysis-definition-nodes-collection.js [hunk 2]
```diff
@@ -66,6 +66,28 @@ module.exports = Backbone.Collection.extend({
         table_name: d.tableName
       }
     });
-  }
+  },
+
+  nodeReady: function (model) {
```
Comment (javierarce) [inline lib/assets/core/javascripts/cartodb3/data/analysis-definition-nodes-collection.js]: I find the name of this method a bit confusing. I would never imagine what it does just looking at it. I would suggest to add a verb (suggestion: `fetchIfOnTop`, `fetchDataIfOnTop`?)

**Comment by matallo** [comment]
actually the original (and non-existent-yet) PR would be from the branch `11874-issue`, thus I didn't make any further change. Shall I point this to master now @xavijam?

**Comment by xavijam** [comment]
Wait, let me add several comments too.

**Comment by javierarce** [comment]
> actually the original (and non-existent-yet) PR would be from the branch 11874-issue, thus I didn't make any further change. Shall I point this to master now @xavijam?

Oh, sorry, I didn't notice that it wasn't pointing to master.

File: app/views/admin/organization_users/_form.html.erb (hunk 1 above)
Comment (xavijam) [inline app/views/admin/organization_users/_form.html.erb]: ?

File: lib/assets/core/javascripts/cartodb3/data/analysis-definition-node-model.js [hunk 3]
```diff
@@ -63,6 +63,10 @@ module.exports = Backbone.Model.extend({
     }
   },
 
+  nodeReady: function () {
```
Comment (xavijam) [inline lib/assets/core/javascripts/cartodb3/data/analysis-definition-node-model.js]: I would not invoke a collection method with a node method too. Why don't you like an event?

Comment (xavijam) [inline lib/assets/core/javascripts/cartodb3/data/analysis-definition-node-model.js]: And let's not use ready for this purpose, that's already used to know if an analysis has finished...

File: lib/assets/core/javascripts/cartodb3/data/analysis-definition-nodes-collection.js (hunk 2 above)
Comment (xavijam) [inline lib/assets/core/javascripts/cartodb3/data/analysis-definition-nodes-collection.js]: Already dixit, node ready is already used and it is confusing.

File: app/views/admin/organization_users/_form.html.erb (hunk 1 above)
Comment (matallo) [inline app/views/admin/organization_users/_form.html.erb]: seems derived from not having `11874-issue` updated from `master` now they are

File: lib/assets/core/javascripts/cartodb3/deep-insights-integrations.js [hunk 4]
```diff
@@ -368,7 +367,7 @@ F.prototype._tryToSetupDefinitionNodeSync = function (m) {
 
       if (status === 'ready') {
         if (!m.__initialization) {
-          m.fetchQueryObjects();
+          m.nodeReady();
```
Comment (xavijam) [inline lib/assets/core/javascripts/cartodb3/deep-insights-integrations.js]: Same as above.

File: lib/assets/core/javascripts/cartodb3/data/analysis-definition-nodes-collection.js (hunk 2 above)
Comment (xavijam) [inline lib/assets/core/javascripts/cartodb3/data/analysis-definition-nodes-collection.js]: Agree with @javierarce

**Comment by matallo** [comment]
made the changes, and moved the method (now renamed) to an event

**Comment by matallo** [comment]
updated master, and now PR points to `master` as requested

**Comment by xavijam** [comment]
Could we change the title and the description of this PR please. Because now it has everything related to the ticket.

File: lib/assets/core/javascripts/cartodb3/editor/layers/layer-content-views/data/data-content-view.js [hunk 5]
```diff
@@ -7,6 +7,7 @@ var dataErrorTemplate = require('./data-content-error.tpl');
 var actionErrorTemplate = require('../../sql-error-action.tpl');
 var QuerySanity = require('../../query-sanity-check');
 var OverlayView = require('../../../components/overlay/overlay-view');
+var checkAndBuildOpts = require('../../../../helpers/required-opts');
 
 var renderLoading = true;
```
Comment (javierarce) [inline lib/assets/core/javascripts/cartodb3/editor/layers/layer-content-views/data/data-content-view.js]: I know this was here before, but does it make sense to use a global var? Wouldn't it be better to use `this._renderLoading` or as a field inside of a model?

Comment (matallo) [inline lib/assets/core/javascripts/cartodb3/editor/layers/layer-content-views/data/data-content-view.js]: actually I think it makes sense to have it in the `modelView` I'll proceed if you're ok with this

Comment (javierarce) [inline lib/assets/core/javascripts/cartodb3/editor/layers/layer-content-views/data/data-content-view.js]: +1

File: lib/assets/core/javascripts/cartodb3/deep-insights-integrations.js [hunk 6]
```diff
@@ -110,7 +110,7 @@ var F = function (opts) {
   this._vis().on('change:error', this._visErrorChange, this);
 
   var saveStateDefinition = _.debounce(this._saveStateDefinition.bind(this), 500);
-  this._diDashboard.onStateChanged(saveStateDefinition);
+  this._diDashboard.onStateChanged(saveStateDefinition, true);
```
Comment (xavijam) [inline lib/assets/core/javascripts/cartodb3/deep-insights-integrations.js]: This `true` is not necessary. I feel I added in my branch but it is not necessary anymore.

**Comment by javierarce** [comment]
I think this is good to go… what do you think, @xavijam?

**Comment by xavijam** [comment]
I feel It can go to acceptance!

**Comment by xavijam** [comment]
Deploying to ded07.

File: app/assets/stylesheets/editor-3/form-placeholder.css.scss [hunk 7]
```diff
@@ -12,6 +12,9 @@ $br: 2px;
   position: relative;
   margin-bottom: 30px;
 }
+.FormPlaceholder-paragraphLoader {
```
Comment (xavijam) [inline app/assets/stylesheets/editor-3/form-placeholder.css.scss]: Uhm? ^^^

Comment (matallo) [inline app/assets/stylesheets/editor-3/form-placeholder.css.scss]: https://github.com/CartoDB/cartodb/blame/8fc88fee00742c6c48c1cbc260d1cec34e93d4b5/app/assets/stylesheets/editor-3/form-placeholder.css.scss#L15

🤷

**Comment by xavijam** [comment]
Something weird is happening with style, data and rest of layer panels. They don't get loaded when widgets are filtered. I'll check at length tomorrow.

File: app/assets/stylesheets/editor-3/form-placeholder.css.scss (hunk 7 above)
Comment (xavijam) [inline app/assets/stylesheets/editor-3/form-placeholder.css.scss]: :joy: sorry.

**Pull request merged**